src/supervisor.rs

```diff
@@ -0,0 +1,614 @@
+    pub failing_sibling_termination_error: Option<TerminationError>,
+}
+
+#[derive(Debug)]
+pub struct TerminationError(HashMap<String, Arc<worker::TerminationError>>);
+
+pub struct Spec {
+    pub name: String,
+    pub ev_notifier: EventNotifier,
+    pub order: Order,
+    pub strategy: Strategy,
+    pub children: Vec<worker::Spec>,
+}
+
+pub struct Supervisor {
+    pub runtime_name: String,
+    pub children: Vec<Worker>,
+
+    pub name: String,
+    pub ev_notifier: EventNotifier,
+    pub order: Order,
+    pub strategy: Strategy,
+}
+
+fn ordered_children_on_start(
+    order: &Order,
+    children_specs: Vec<worker::Spec>,
+) -> Box<dyn Iterator<Item = worker::Spec>> {
+    match order {
+        Order::LeftToRight => Box::new(children_specs.into_iter()),
+        Order::RightToLeft => Box::new(children_specs.into_iter().rev()),
+    }
+}
+
+fn ordered_children_on_termination(
+    order: &Order,
+    children: Vec<Worker>,
+) -> Box<dyn Iterator<Item = Worker>> {
+    match order {
+        Order::LeftToRight => Box::new(children.into_iter().rev()),
+        Order::RightToLeft => Box::new(children.into_iter()),
+    }
+}
+
+// terminate_prev_siblings gets called when a child in a supervision tree fails
+// to start, and all it's previous siblings need to get terminated.
+async fn terminate_prev_siblings<'a, 'b, 'c>(
+    ev_notifier: &mut EventNotifier,
+    failed_child_spec: worker::Spec,
+    failed_child_start_err: Arc<worker::StartError>,
+    pending_children_specs: Vec<worker::Spec>,
+    started_children: Vec<Worker>,
+) -> (Vec<worker::Spec>, StartError) {
+    let prev_sibling_termination_result =
+        terminate_children(ev_notifier, started_children.into_iter().rev()).await;
+
+    let (prev_specs, failing_sibling_termination_error) = match prev_sibling_termination_result {
+        Err((prev_specs, sibling_err)) => (prev_specs, Some(sibling_err)),
+        Ok(prev_specs) => (prev_specs, None),
+    };
+
+    let start_err = StartError {
+        failing_child_runtime_name: "".to_owned(),
+        failing_child_error: failed_child_start_err,
+        failing_sibling_termination_error,
+    };
+
+    // regenerate the original vector of worker
+    // specs so that the parent supervisor can try
+    // again if it is able to do it
+    let children_spec: Vec<worker::Spec> = prev_specs
+        .into_iter()
+        .chain(vec![failed_child_spec].into_iter())
+        .chain(pending_children_specs.into_iter())
+        .collect();
+
+    (children_spec, start_err)
+}
+
+// StartResult is an internal state machine used when starting workers in a
+// supervisor
+enum StartResult {
+    SuccessfulStart {
+        started_workers: Vec<Worker>,
+        pending_count: usize,
+    },
+    FailedStart {
+        prev_started_workers: Vec<Worker>,
+        failed_worker_spec: worker::Spec,
+        start_error: Arc<worker::StartError>,
+        pending_workers: Vec<worker::Spec>,
+    },
+}
```

Comment:
This type and it's usage feels a bit weird to me. I see what's happening, and get the accumulator/state machine nature of it, but it strikes me as very unidiomatic.  Maybe it's because it's essentially playing the same role as a `Result` without a `Result`.

Reply:
I hate this type personally, coming from Haskell land, we would have recursion here and we would be able to have the vector of pending children consumed up to a point, in Rust, I'm forced to move the vector which makes me loose the unused children.

I note that we may be able to get away with it via `&mut` on a different comment, I'll comment about that there.